# DS 14. Maxfooling2D & Conv2D를 이용한 풀이

> "작성완료"

- toc: true
- branch: master
- badges: true
- comments: true
- [python, Data Science]

---
# Data Science
- lenture: Data Science_11-1nd week of lectures.
- lenture date: 2022-05-11
- lecturer: Guebin choi
- study date: 2022-05-12
- author: Kione kim
---

In [1]:
import tensorflow as tf 
import tensorflow.experimental.numpy as tnp

In [2]:
tnp.experimental_enable_numpy_behavior()

In [3]:
import matplotlib.pyplot as plt

## review

`-` 데이터

In [8]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

In [9]:
y_train.shape

(60000,)

In [10]:
X= tf.constant(x_train.reshape(-1,28,28,1),dtype=tf.float64)
y= tf.keras.utils.to_categorical(y_train)
XX= tf.constant(x_test.reshape(-1,28,28,1),dtype=tf.float64)
yy= tf.keras.utils.to_categorical(y_test)

`-` net1

In [28]:
tf.random.set_seed(43052)
net1= tf.keras.Sequential()
net1.add(tf.keras.layers.Flatten())
net1.add(tf.keras.layers.Dense(30, activation='relu'))
net1.add(tf.keras.layers.Dense(10, activation='softmax'))
net1.compile(loss=tf.keras.losses.categorical_crossentropy, optimizer= 'adam', metrics= ['accuracy'])
net1.fit(X,y,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 9s 4ms/step - loss: 2.5388 - accuracy: 0.4048
Epoch 2/5
1875/1875 [==============================] - 7s 4ms/step - loss: 1.2080 - accuracy: 0.5134
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 1.0115 - accuracy: 0.5905
Epoch 4/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.9389 - accuracy: 0.6169
Epoch 5/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.8639 - accuracy: 0.6457


In [29]:
net1.evaluate(XX,yy)

313/313 [==============================] - 1s 3ms/step - loss: 0.8616 - accuracy: 0.6555


[0.8615648150444031, 0.6554999947547913]

`-` net2

In [30]:
tf.random.set_seed(43052)
net2= tf.keras.Sequential()
net2.add(tf.keras.layers.Flatten())
net2.add(tf.keras.layers.Dense(500, activation= 'relu'))
net2.add(tf.keras.layers.Dense(500, activation= 'relu'))
net2.add(tf.keras.layers.Dense(500, activation= 'relu'))
net2.add(tf.keras.layers.Dense(500, activation= 'relu'))
net2.add(tf.keras.layers.Dense(10, activation= 'softmax'))
net2.compile(loss=tf.keras.losses.categorical_crossentropy, optimizer= 'adam', metrics=['accuracy'])
net2.fit(X,y,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 46s 24ms/step - loss: 1.1121 - accuracy: 0.7888
Epoch 2/5
1875/1875 [==============================] - 47s 25ms/step - loss: 0.4588 - accuracy: 0.8368
Epoch 3/5
1875/1875 [==============================] - 38s 20ms/step - loss: 0.4150 - accuracy: 0.8538
Epoch 4/5
1875/1875 [==============================] - 38s 20ms/step - loss: 0.3933 - accuracy: 0.8618
Epoch 5/5
1875/1875 [==============================] - 37s 20ms/step - loss: 0.3712 - accuracy: 0.8687


In [31]:
net2.evaluate(XX,yy)

313/313 [==============================] - 3s 9ms/step - loss: 0.4488 - accuracy: 0.8496


[0.4488353431224823, 0.8496000170707703]

- DNN 아키텍처로는 한계가 있음
- 다른 방안이 필요함!

## Maxfooling 2D, Conv2D 이용한 풀이

`-` net3

In [11]:
tf.random.set_seed(43052)
net3= tf.keras.Sequential()
net3.add(tf.keras.layers.Conv2D(30,(2,2),activation='relu')) # layer1
net3.add(tf.keras.layers.MaxPool2D()) # layer2
net3.add(tf.keras.layers.Conv2D(30,(2,2),activation='relu')) # layer3
net3.add(tf.keras.layers.MaxPool2D()) # layer4
net3.add(tf.keras.layers.Flatten()) # layer5
net3.add(tf.keras.layers.Dense(10,activation='softmax')) # layer6
net3.compile(loss=tf.keras.losses.categorical_crossentropy, optimizer='adam', metrics=['accuracy'])
net3.fit(X,y,epochs=5)

Epoch 1/5
1875/1875 [==============================] - 40s 20ms/step - loss: 0.7706 - accuracy: 0.8093
Epoch 2/5
1875/1875 [==============================] - 31s 17ms/step - loss: 0.3572 - accuracy: 0.8723
Epoch 3/5
1875/1875 [==============================] - 36s 19ms/step - loss: 0.3161 - accuracy: 0.8868
Epoch 4/5
1875/1875 [==============================] - 33s 18ms/step - loss: 0.2931 - accuracy: 0.8942
Epoch 5/5
1875/1875 [==============================] - 33s 17ms/step - loss: 0.2757 - accuracy: 0.8995


- 기존 DNN 모형에서는 Flatten( )을 이용하여 N×p의 형태로 만들어 준 후 layer들을 추가하여 사용하였는데,
- 위 모형에서는 Conv 2d, Maxpool 2d, Conv 2d, Maxpool 2d layer를 사용한 다음 Flatten()을 사용하였다.
- 즉, 위 모형에서 Flatten()을 사용하기 전까지의 형태는 이미지 자료가 입력으로 들어가서 이미지 자료가 출력으로 나온 것이다.

In [12]:
net3.evaluate(XX,yy)

313/313 [==============================] - 2s 6ms/step - loss: 0.3535 - accuracy: 0.8716


[0.35350102186203003, 0.8715999722480774]

- 기존의 DNN 모형으로 분류한 것보다 더 높은 accuracy가 나왔다! -> DNN의 한계를 가볍게 돌파!!

### net3의 구조(네트워크 뜯어보기)

In [17]:
net3.layers

In [18]:
cv1, mp1, cv2, mp2, flttn, d1 = net3.layers # 언패킹

In [26]:
print(XX.shape)
print(cv1(XX).shape)
print(mp1(cv1(XX)).shape)
print(cv2(mp1(cv1(XX))).shape)
print(mp2(cv2(mp1(cv1(XX)))).shape)
print(flttn(mp2(cv2(mp1(cv1(XX))))).shape)
print(d1(flttn(mp2(cv2(mp1(cv1(XX)))))).shape)

(10000, 28, 28, 1)
(10000, 27, 27, 30)
(10000, 13, 13, 30)
(10000, 12, 12, 30)
(10000, 6, 6, 30)
(10000, 1080)
(10000, 10)


- shape을 잘 따지는 게 중요하다! 동작원리를 이해하는데 도움!!!
- Conv2D를 통과시키니 이미지 사이즈는 기존 사이즈보다 가로, 세로 하나씩 작아졌고 채널은 30이 되었음.
- Maxpool2D를 통과시키니 이미지 사이즈는 기존 사이즈(27×27)보다 약 1/2 작아진 (13×13)이 되었음. (2로 나누고 하나가 남으면 버림)
- 두 번째 Conv2D, Maxpool2D는 마찬가지이다.
- Flatten()을 해주면 obs를 제외한 나머지를 shrinke해줌!(6×6×30 -> 1080)
- Flatten은 (obs, 가로픽셀, 세로픽셀, 채널)을 가로픽셀 × 세로픽셀 × 채널 = p 로 해주는 것임!

- 의문: 흠,, 근데 이 모형 좀 복잡한 거 아닌가?

### 파라메터 수 비교

In [32]:
net1.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_2 (Flatten)         (None, 784)               0         
                                                                 
 dense_2 (Dense)             (None, 30)                23550     
                                                                 
 dense_3 (Dense)             (None, 10)                310       
                                                                 
Total params: 23,860
Trainable params: 23,860
Non-trainable params: 0
_________________________________________________________________


In [33]:
net2.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_3 (Flatten)         (None, 784)               0         
                                                                 
 dense_4 (Dense)             (None, 500)               392500    
                                                                 
 dense_5 (Dense)             (None, 500)               250500    
                                                                 
 dense_6 (Dense)             (None, 500)               250500    
                                                                 
 dense_7 (Dense)             (None, 500)               250500    
                                                                 
 dense_8 (Dense)             (None, 10)                5010      
                                                                 
Total params: 1,149,010
Trainable params: 1,149,010
No

In [34]:
net3.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 27, 27, 30)        150       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 30)       0         
 2D)                                                             
                                                                 
 conv2d_3 (Conv2D)           (None, 12, 12, 30)        3630      
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 6, 6, 30)         0         
 2D)                                                             
                                                                 
 flatten_1 (Flatten)         (None, 1080)              0         
                                                                 
 dense_1 (Dense)             (None, 10)               

1. net1의 파라메터: 23,860
2. net2의 파라메터: 149,010
3. net3의 파라메터: 14,590

In [36]:
14590/23860

0.6114836546521375

In [37]:
14590/149010

0.09791289175223139

- net3의 파라메터수는 net1의 3/5 수준이고 net2의 1/10 수준이다.